### PR TITLE
Resampling in mini-batches

### PR DESCRIPTION
**Pull request by mzient**
Resampling operators (Resize, RandomResizedCrop) can now process images in mini-batches. Mini-batch must be smaller than batch size - otherwise it's ignored.

Changes:
- convert resampling `Params` from `vector` to `span`
- add multiple kernel data objects to GPU kernel
- bug fixes

### CODE REVIEW

File: dali/pipeline/operators/resize/resize_base.cc [hunk 1]
```diff
@@ -71,31 +73,89 @@ ResamplingFilterAttr::ResamplingFilterAttr(const OpSpec &spec) {
...
-      view<const uint8_t, 3>(input),
-      resample_params_);
+  auto &alloc = GetGPUScratchAlloc();
 
-  kdata.scratch_alloc.Reserve(kdata.requirements.scratch_sizes);
+  auto in_view = view<const uint8_t, 3>(input);
+  SubdivideInput(in_view);
+
+  decltype(kernels::KernelRequirements::scratch_sizes) scratch_sizes;
+  for (auto &size : scratch_sizes)
```
Comment (jantonguirao) [inline dali/pipeline/operators/resize/resize_base.cc]: isn't it empty at this point?

File: dali/kernels/util.h [hunk 2]
```diff
@@ -38,11 +38,34 @@ size_t size(const T (&a)[N]) {
   return N;
 }
 
+
```
Comment (Kh4L) [inline dali/kernels/util.h]: Remove extra \n?

File: dali/pipeline/operators/resize/resize_base.cc [hunk 3]
```diff
@@ -104,13 +164,24 @@ void ResizeBase::Initialize(int num_threads) {
   resample_params_.resize(num_threads);
 }
 
-void ResizeBase::InitializeGPU() {
-  kernel_data_.resize(1);
-  auto &kdata = GetKernelData();
+void ResizeBase::InitializeGPU(int batch_size, int mini_batch_size) {
+  DALI_ENFORCE(batch_size > 0, "Batch size must be positive");
+  DALI_ENFORCE(mini_batch_size > 0, "Mini-batch size must be positive");
```
Comment (jantonguirao) [inline dali/pipeline/operators/resize/resize_base.cc]: shouldn't you enforce that `mini_batch_size` is actually smaller than `batch_size`?

File: dali/pipeline/operators/resize/resize_base.cc (hunk 1 above)
Comment (Kh4L) [reply dali/pipeline/operators/resize/resize_base.cc]: No its not, `scratch_sizes` type is `std::array<size_t, (size_t)AllocType::Count>`

**Comment by mzient** [comment]
Build passed: 695148

File: dali/pipeline/operators/resize/resize_base.cc (hunk 3 above)
Comment (mzient) [reply dali/pipeline/operators/resize/resize_base.cc]: This is a bit of a guideline. E.g. if you specify minibatch of 32 and batch of 208, then you get some minibatches of 29, some of 30 (and none 32). If batch_size < minbatch_size, then you get just 1 minibatch.

File: dali/pipeline/operators/resize/resize_base.cc (hunk 1 above)
Comment (mzient) [reply dali/pipeline/operators/resize/resize_base.cc]: It's std::array - it's uninitialized.

File: dali/kernels/span.h [hunk 4]
```diff
@@ -152,6 +153,27 @@ constexpr span<T, Extent> make_span(T *data) { return { data }; }
...
+template <typename Collection>
+auto make_span(Collection &c)->decltype(make_span(c.data(), c.size())) {
+  return make_span(c.data(), c.size());
+}
+
+template <typename T, size_t N>
+constexpr span<T, N> make_span(std::array<T, N> &a) {
+  return { a.data() };
+}
+
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/span.h]: Done.

**Comment by mzient** [comment]
Build passed: 695238

**Pull request merged**